Employee validation

RomanZubariev opened the pull request:
- added EmployeeBiethDate constraint annotation to apply EmployeeBirthDateValidator that checks for the birth date to be at least 18 years ago.
- added min and max constraints to Employee.employeeId and Employee.departmentId
- EmployeeController now checks if id and requested Body are @Valid
- GlobalExceptionHandler now checks for ConstraintViolationException (is thrown when parameters in controller are not valid)
  handling MethodArgumentNotValidException  was redefined to show the validation messages(the default method in ResponseEntityExceptionHandler doesn't return any information about validation issues)
- tested Employees were changed.

RomanZubariev1990 committed:
Employee validation

- added EmployeeBiethDate constraint annotation to apply EmployeeBirthDateValidator that checks for the birth date to be at least 18 years ago.
- added min and max constraints to Employee.employeeId and Employee.departmentId
- EmployeeController now checks if id and requested Body are @Valid
- GlobalExceptionHandler now checks for ConstraintViolationException (is thrown when parameters in controller are not valid)
  handling MethodArgumentNotValidException  was redefined to show the validation messages(the default method in ResponseEntityExceptionHandler doesn't return any information about validation issues)
- tested Employees were changed.

HalinaPichukha commented on src/main/java/com/mastery/java/task/dto/EmployeeBirthDate.java:
```diff
@@ -0,0 +1,24 @@
+
+@Target({FIELD})
+@Retention(RUNTIME)
+@Constraint(validatedBy = EmployeeBirthDateValidator.class)
```
Constraint уже включает в себя аннотации @Retention(RUNTIME) и @Documented.

RomanZubariev replied:
Я не уверен что мета-аннотации так работают. Они действуют на аннотацию @Constraint , а не на ту кастомную которую я задаю. По крайней мере в примере с docs.oracle.com эти мета-аннотации есть, и без retention валидация возраста не работает.

pavel-tsikhanovich replied:
Можно ещё посмотреть, как у Spring они находятся в каком-нибудь Controller/RestController или Component/Другие бины

HalinaPichukha commented on src/main/java/com/mastery/java/task/dto/EmployeeBirthDateValidator.java:
```diff
@@ -0,0 +1,14 @@
+
+  @Override
+  public boolean isValid(LocalDate date, ConstraintValidatorContext context) {
+    return date.isBefore(LocalDate.now().minusYears(18));
```
Эту логику было бы хорошо покрыть тестами)

RomanZubariev replied:
Написал юнит тест для этого метода. Оказалось что метод работает не так как ожидается в день рождения (наверное), заменил на ! + isAfter.

RomanZubariev1990 committed:
Validation tests

- Added tests for birth date validator
- Employee is now valid even if they turn 18 today.

RomanZubariev1990 committed:
Formatting
